### ISSUE TITLE
Youtube: Segment number incorrect sometimes

### ISSUE THREAD

**Issue by lekoOwO** [streamlink/streamlink#3496]
## Plugin Issue

- [x] This is a plugin issue and I have read the contribution guidelines.
- [x] I am using the latest development version from the master branch.

### Description
When downloading youtube live streams, the segment number seems not to be correct sometimes.
As far as I know, the segment number means "the second in stream", i.e.: segment 60 means the 60th second in the video/livestream. Did I get it right? (It applies to many of my recordings)
But sometimes the segment number is not correct, mostly when the record does not start from the very beginning of the live stream.

### Reproduction steps / Explicit stream URLs to test
Can't be reproduced everytime, just record a random youtube livestream and it may happen.

### Log output

```
[04:04:18.803008][cli][info] Available streams: 144p (worst), 240p, 360p, 480p, 720p, 1080p (best)
[04:04:18.803147][cli][info] Opening stream: 1080p (hls)
[04:04:18.803606][stream.hls][debug] Reloading playlist
[04:04:18.836605][stream.hls][debug] First Sequence: 126; Last Sequence: 128
[04:04:18.836699][stream.hls][debug] Start offset: 0; Duration: None; Start Sequence: 126; End Sequence: None
[04:04:18.836960][stream.hls][debug] Adding segment 126 to queue
[04:04:18.837041][cli][debug] Pre-buffering 8192 bytes
[04:04:18.837407][stream.hls][debug] Adding segment 127 to queue
[04:04:18.841539][stream.hls][debug] Adding segment 128 to queue
[04:04:19.150398][stream.hls][debug] Download of segment 126 complete
....
```

### Additional comments, etc.
The segment 126 in log is actually segment 272.

[Love Streamlink? Please consider supporting our collective. Thanks!](https://opencollective.com/streamlink/donate)

**Comment by beardypig** [comment]
Seems like you are conflating segments and seconds. A segment does not have a fixed duration, there is a target duration set in the playlist, that duration is specified by whomever is generating the playlist (in this case YouTube, how that duration is picked is anyones guess though). The segment length is likely to remain constant throughout the stream, so you can normally relate the duration of the stream to the segment number. 

Are you trying to use the segment number for something?

**Comment by lekoOwO** [comment]
> Are you trying to use the segment number for something?

Yes, I'm trying to get the start and end time in the original livestream which streamlink has downloaded from the log file.

**Comment by bastimeyer** [comment]
> I'm trying to get the start and end time in the original livestream which streamlink has downloaded from the log file

You are looking for the [`EXT-X-PROGRAM-DATE-TIME`](https://tools.ietf.org/html/rfc8216#section-4.3.2.6) metadata of the first and last segment. This HLS tag is optional and it can either be completely missing or missing on consecutive segments in the HLS playlist (after the first segment or a discontinuity). If there's at least one available, the date time of the following segments can be calculated from the segment durations.

The start time of the output stream could be logged by Streamlink if it's available, same as the end time. Maybe on the debug logging level, but I could also see it being useful on the info level.

**Comment by lekoOwO** [comment]
> The start time of the output stream could be logged by Streamlink if it's available, same as the end time. Maybe on the debug logging level, but I could also see it being useful on the info level.

hmm, I did not see it in the log file, perhaps Youtube does not provide it?

**Comment by bastimeyer** [comment]
I said it would be useful as an addition. I didn't say that it's currently implemented.

**Comment by beardypig** [comment]
> perhaps Youtube does not provide it?

They don't appear to provide `EXT-X-PROGRAM-DATE-TIME` in any case.

**Comment by lekoOwO** [comment]
> > perhaps Youtube does not provide it?
> 
> They don't appear to provide `EXT-X-PROGRAM-DATE-TIME` in any case.

Well, I tried downloading youtube live stream using youtube-dl, and I see the following logs:
```
[hls @ 00000218cbaff340] Skip ('#EXT-X-VERSION:3')
[hls @ 00000218cbaff340] Skip ('#EXT-X-DISCONTINUITY-SEQUENCE:1')
[hls @ 00000218cbaff340] Skip ('#EXT-X-PROGRAM-DATE-TIME:2021-01-23T03:48:51.864+00:00')
[https @ 00000218cbf76ac0] Opening 'https://r5--.....
[hls @ 00000218cbaff340] Skip ('#EXT-X-VERSION:3')
[hls @ 00000218cbaff340] Skip ('#EXT-X-DISCONTINUITY-SEQUENCE:1')
[hls @ 00000218cbaff340] Skip ('#EXT-X-PROGRAM-DATE-TIME:2021-01-23T03:48:52.864+00:00')
[https @ 00000218cbf76ac0] Opening 'https://r5---sn-ipoxu- .....
```
It seems that it's providing EXT-X-PROGRAM-DATE-TIME every second?

**Comment by beardypig** [comment]
Interesting, maybe it depends on the stream, can you post the URL for that one?

**Comment by lekoOwO** [comment]
> Interesting, maybe it depends on the stream, can you post the URL for that one?

[link](https://www.youtube.com/watch?v=qTwHCC4stu8)

In my experience, I see it every time when using `youtube-dl`... hmm
